### ISSUE TITLE
LSTM layer support

### ISSUE THREAD

**Issue by anbac** [Dobiasd/frugally-deep#8]
It would be very nice to also support LSTM layers

**Comment by Dobiasd** [comment]
Totally agree. Are you interested in implementing it? :-)

**Comment by rcshubhadeep** [comment]
I was about to write about the same and then found this issue thread. I think it will be a great addition. If you want I can try my luck. DId not do C++ in a while but would be great to go back.

**Comment by Dobiasd** [comment]
@rcshubhadeep That would be awesome! Once the algorithm works, we can adjust the C++ style afterwards.

So if you'd like to start implementing, here are the steps that are needed:

- use the layer type in question in `get_test_model_small` in `keras_export/generate_test_models.py`
- add an export function for the layer (if needed) to `keras_export/convert_model.py`
- add the new layer class to `include/fdeep/layers` (actual implementation)
- add a create function (deserialization from JSON) to `include/fdeep/import_model.hpp`
- also in `include/fdeep/import_model.hpp`, extend the `unordered_map` `creators` in `create_layer` to also dispatch to your new create function.

When you have all this, your new layer type will automatically be tested (i.e. its results compared with the Keras implementation) in the unit tests. You can run the tests locally as described in `README.md` or by creating a pull request here in this repo and letting Travis CI do it.

If any questions arise, just let me know. :)

**Comment by rcshubhadeep** [comment]
Here to say, I am sorry but a newborn kid and the pressure of my present job presently preventing me from taking up this venture. Would have loved to

**Comment by Dobiasd** [comment]
No problem at all. Having kids is one of the few things that is even more awesome than implementing deep-learning layers. :wink: So I totally understand.

**Comment by chammika** [comment]
@Dobiasd  I briefly looked into implementing SimpleRNN recurrent layer, weight layout of model seems simple and exporting them wouldn't be hard. I did the forward pass within python using numpy and it agrees with the Keras predictions.
https://gist.github.com/chammika/0448fbb0e96d1365326721137ff86da6

Next steps would be to follow your outline above to have a functional layer. I am not clear yet how to handle the extra time dimension in the data though...

Regarding recurrent layers in general, input shape has another dimension for time ie. `seq_length` How would you envision it fit in to the fdeep ? should we define `tensor4` and `shape4` with `(seq_length, depth, height, width)` 
And if you count the batch_size dimension (which is needed to implement stateful RNN layers)  it's a 5D tensor. Would be great if you can give some ideas on it.

**Comment by Dobiasd** [comment]
@chammika Wow, good work on implementing it in Python already!

Yes, it seems like we need a `Tensor4`, but I'm not yet sure if it makes sense to let all fdeep-layers just work with `Tensor4` instead of `Tensor3`to unify things. The `seq_length` dimension might be confused with a batch-size dimension. What do you think?

Do stateful RNN layers need a fifth dimension (`batch_size`) also when leaving out training and just implementing forward passes?

**Comment by chammika** [comment]
If we are to add `Tensor4` it only apply to the following to cases to my knowledge:
- inputs of RNN layers
- outputs of RNN layers if the `return_sequences=True` (next layer must be a RNN layer)

When `return_sequences=False` of a RNN layer output is `Tensor3` which matches inputs with the non-recurrent layers. To illustrate this I updated the above example code to have two SimpleRNN layers and Dense layer and simplified the calculations a bit.
https://gist.github.com/chammika/0448fbb0e96d1365326721137ff86da6

Note that First SimpleRNN return a sequence which can be fed to a RNN where as second one does not/ it cannot if we are to connect a Dense layer next. I could replicate the Kras predictions in numpy as before.

> but I'm not yet sure if it makes sense to let all fdeep-layers just work with Tensor4 instead of Tensor3to unify things

There is no need Only the above cases will have `Tensor4`

> Do stateful RNN layers need a fifth dimension (batch_size) also when leaving out training and just implementing forward passes?

I am not clear about this yet. According to above example it doesn't. Changing the `stateful = True` or `False` Keras will give same predictions. So I am guessing it's only option during the training ? According to this article however, we should have a difference

https://fairyonice.github.io/Understand-Keras's-RNN-behind-the-scenes-with-a-sin-wave-example.html

Enlighten me if you figure this out :smiley:

**Comment by Dobiasd** [comment]
> There is no need Only the above cases will have Tensor4

Yeah, but if one layer type takes `Tensor4` and the other layer type take `Tensor3`, what type should [`model::predict`](https://github.com/Dobiasd/frugally-deep/blob/4863787d095478b5fc1617e75d9305d51ed23da9/include/fdeep/model.hpp#L23) (and [`layer::apply`](https://github.com/Dobiasd/frugally-deep/blob/4863787d095478b5fc1617e75d9305d51ed23da9/include/fdeep/layers/layer.hpp#L53)) take?

I guess we need some kind of unification, or do you see a loophole somewhere? :slightly_smiling_face:

> So I am guessing it's only option during the training ? According to this article however, we should have a difference

:confused:

**Comment by Dobiasd** [comment]
Leaving the stateful question aside, perhaps there is no problem with "tensor4". Currently `model::predict` and `layer::apply` take `fdeep::tensor3s` (i.e., `std::vector<fplus::tensor3>`, because that's just how models are implemented in Keras. They can take multiple input tensors and some layer types also take multiple tensors as input (e.g., [merge layers](https://keras.io/layers/merge/)). Layers like `conv2d` of course only can work when this [tensor3 vector has size 1](https://github.com/Dobiasd/frugally-deep/blob/4863787d095478b5fc1617e75d9305d51ed23da9/include/fdeep/layers/conv_2d_layer.hpp#L53). Perhaps LSTM is just like for example the [`concatenate layer`](https://github.com/Dobiasd/frugally-deep/blob/4863787d095478b5fc1617e75d9305d51ed23da9/include/fdeep/layers/concatenate_layer.hpp) taking a tensor3 vector with more than 1 element. In that case we could just add LSTM without changing the overall structure and types of the lib. What do you think?

**Comment by chammika-become** [comment]
Yes that would be the best solution keep the current layer api intact. In that case `size(tensor3s) == seq_len`. According to an answer on below question

https://stackoverflow.com/questions/42763928/how-to-use-model-reset-states-in-keras

> stateful=True is usually used when you want to treat consecutive batches as consequtive inputs. In this case model is treating consequtive batches the same as it were in the same batch.

If we flatten batches in to one sequence we can have the same effect as a having the stateful predictions. In that case we might want to update the api to

`apply(const tensor3s& inputs, batch_size=1)`

Pass the appropriate `batch_size` so that the outputs will be generated from the layer at `batch_size` intervals.
Since I haven't seen any difference in stateful predictions form my simulations of Kera models,  I would first implement layer without `stateful` and see if the predictions really need this extra `batch_size` parameter to work.

**Comment by Dobiasd** [comment]
Sounds good. :+1:

One question though: You mean changing the interface of [`fdeep::layer::apply`](https://github.com/Dobiasd/frugally-deep/blob/4863787d095478b5fc1617e75d9305d51ed23da9/include/fdeep/layers/layer.hpp#L53
) from `virtual tensor3s apply(const tensor3s& input) const final` to `virtual tensor3s apply(const tensor3s& input, batch_size=1) const final`?

Since frugally-deep does not support any kind of batching (not counting [parallel predictions](https://github.com/Dobiasd/frugally-deep/blob/4863787d095478b5fc1617e75d9305d51ed23da9/include/fdeep/model.hpp#L40)) I'd like to keep the word batch out. Do you think it is possible to leave the interface as it, and just assume (or calculate) the batch size if needed inside the LSTM-layer implementation?

**Comment by chammika-become** [comment]
It's possible to keep the interface as it is for `stateful=False` layers. It **might** be necessary to change the signature of the interface as above. Yes you are right about my suggestion it's virtual `layer::apply` method. 
Still, I am not sure if it's necessary because I am not fully clear on stateful predictions, sorry for confusing you with suggestions that fully don't understand.
I will first implement SimpleRNN layer (with `stateful=Fale`) and see if it can predict stateful layer without modifying the interface.

**Comment by Dobiasd** [comment]
Sounds like a very good plan. :+1:

Let me know if you have any questions regarding the integrations or if I can be of any other kind of help. :slightly_smiling_face:

**Comment by n-Guard** [comment]
@Dobiasd First of all: keep up the good work, I think this project is really useful!

I recently got my first job at an audio software company and it's my task to implement the `LSTM`, `TimeDistributed`, and `Bidirectional` layers for production.

I took a shot at implementing the `lstm_layer` class with the inner function that computes the LSTM output, but only for my special use case where I basically just have a 2-dimensional input with (`seq_length`, `1D_data`). The third dimension (first in `Tensor3`) is ignored right now because it seemed to me that this would be a little bit harder to implement since we don't work with Eigen-unsupported where n-dimensional matrices can be used.

For testing, I manually copied the weights from the keras model and fed them into the function. Output values are reasonably close to keras output.
I know the implementation is probably far from ideal because I'm relatively new at C++  but maybe it's a starting point. 

Actual implementation: https://gist.github.com/n-Guard/50a64f4ab837b06777263758b15e6118
Tests: https://gist.github.com/n-Guard/afe8a288ac54a32880ceb3f1e37d95d8

Next thing would obviously be implementing the model conversion to be able to actually test random cases.

I'm looking forward to get some feedback!

**Comment by Dobiasd** [comment]
@n-Guard Hi, and thanks for the nice feedback and this good contribution. Code cleanup regarding C++ style can be done later. :)

The more important thing from my POV is that we now seem to have two people working on the same feature. Maybe to further improve quality (and to not waste yout time too) a collaboration of you two would be reasonable? @chammika / @chammika-become What do you think?

**Comment by chammika-become** [comment]
@n-Guard Great work! :+1: 
@Dobiasd I have been working on SimpleRNN managed only so far to export weights from model. On the otherhand @n-Guard have already implemented a working version LSTM so let's go ahead integrating his solution.
Only thing I would suggest now is to keep `bias` in `RowMajorMatrixXf` from from the beginning so you don't have to regenerate the `b` here for every prediction.
https://gist.github.com/n-Guard/50a64f4ab837b06777263758b15e6118#file-lstm_layer-hpp-L123

Could you also post a link to the python model that comes up with the same predictions for easy of comparison. I particularly want to see how the `seq_len` of the input vectors is handled. My idea for SimpleRNN is `len(tensor3s) == seq_len` where as in your case it looks like `x_width`

**Comment by Dobiasd** [comment]
OK, cool. @n-Guard now that this part is clear, feel free to open a PR with your work if you like, so we can view it and so that you have automatic tests in the CI etc.

**Comment by Dobiasd** [comment]
> But I think your suggestion with `seq_len` being the first dimension is more consistent, so let's go with that!

`fdeep::layer::apply` takes `tensor3s` (i.e., `vector<tensor3>`). Would it make sense to use the size of this vector as `seq_len` instead of reusing "channels, "height" or "width"?

---

> Btw is there a specific reason why you didn't use Eigen-unsupported for having n-dimensional matrices? I'm interested because tensorflow seems to use it.

I did not yet look into it. The main reason might be historical. For quite a long time `fdeep` did not use Eigen at all. I then only added it to speed up the matrix multiplications. ;)

**Comment by n-Guard** [comment]
> fdeep::layer::apply takes tensor3s (i.e., vector<tensor3>). Would it make sense to use the size of this vector as seq_len instead of reusing "channels, "height" or "width"? 

Ah, so if I understand it correctly you would have a vector with elements representing multiple `tensor3` instances (i.e. `tensor3s`), where each `tensor3` element represents one time step with three dimensions (channels, height, width)? If so, I think this would make sense.

**Comment by Dobiasd** [comment]
Yes, exactly. And to me this sounds more intuitive. On the other hand, I don't know much about LSTMs. So if this poses a problem, feel free to use an alternative approach.

**Comment by n-Guard** [comment]
After doing some research, I don't think there is a problem specific to LSTMs since keras LSTM-layer (and RNN-layers in general) only accept a 3D Tensor as input (`batch_size`, `timesteps`, `input_dim`), so basically we only have 2 dimensions left when leaving out batch dimension. This also implies that we don't need n-dimensional matrices at this point.
So the problem of having more than 3 dimensions only occurs when the input layer is not an RNN-layer.

**Comment by n-Guard** [comment]
Ok, I think we can definitely agree on having `seq_len` as `tensor3s.size()` since time dimension is always first dimension in keras.
In my lastest commit to my PR, I changed the input accordingly but with the feature dimension being `tensor3.height()` because I thought this would be most appropriate (having no channel-dimension), so the input to the LSTM layer would be `(seq_len, 1, n_features, 1)`. 

**But:** Since the `flatten` layer returns a tensor3 with `shape (n, 1, 1)` I think it would be better to have `tensor3.depth()` as the feature dimension because when we have a greater than 2-dim input to an RNN layer (e.g. from a `Conv2D` layer) the input has to be flattened to use in the RNN layer.

**Comment by Dobiasd** [comment]
Good catch! `flatten` (in Keras and in fdeep) only makes the depth dimension != 1. So if it's a normal use case in Keras to do `flatten -> LSTM`, how do the LSTM layers there work with this output if they expect their `seq_len` as "number of tensors" (instead of depth)?

Is it possible for us at all to "just" do it like Keras does, or is this in conflict with how we handle tensors in general?

(It might be that I don't understand the problem fully, so please correct me if needed. :slightly_smiling_face:)

**Comment by n-Guard** [comment]
When an `LSTM` layer is used after a non-RNN layer it might be necessary to use the `TimeDistributed` wrapper to keep the time dimension intact so it won't be affected by the `flatten` operation. So a tensor with shape (`seq_len, channels, height, width`) will then be of shape (`seq_len, channels*height*width`)

> Is it possible for us at all to "just" do it like Keras does, or is this in conflict with how we handle tensors in general?

Right now I don't see why this should be a problem in fdeep.

**Comment by n-Guard** [comment]
> edit: I know it would be way cooler if I could dive deep into LSTM layers together with you. But sadly, currently it's not easy for me to allocate time to understand this layer type, that I've not yet worked with, completely. Sorry for that.

That's ok, it's already super awesome that you provided us with this great library in the first place! 🙂

Speaking of major hacks, I got the implementation with conversion from Keras + test working locally, but I ran into some issues, that's why I didn't update my PR yet. 

First thing is: In Python, when I use an LSTM layer in `get_test_model_small()`, it expects two dimensions (`seq_len`, `n_features`). Then in `convert_model.py` the input dimension is swapped like this: https://github.com/Dobiasd/frugally-deep/blob/master/keras_export/convert_model.py#L63
This leads to my tensor3 being (`n_features`, `1`, `seq_len`) which is a little bit confusing. So in order to get it working with the tests I temporarily set the input and output of the `lstm_layer` class accordingly (not using `tensor3s` elements as timesteps). I don't really understand why it gets swapped anyway with two dimensions. What is the assumption here?

At this point I'm not really sure what's the best testing pipeline in order to support our agreement on using `tensor3s` as input to the `lstm_layer`, so I appreciate any suggestion 🙂.

Side note: It took me a while to realise that if there is a keyword "activation" in the model.json file, a global activation is applied to the output of the layer. Unfortunately, this is wrong in LSTM layers, so I had to make an exception here (just so you know): https://github.com/Dobiasd/frugally-deep/blob/master/include/fdeep/import_model.hpp#L856

**Comment by Dobiasd** [comment]
>  What is the assumption here?

I probably should have added some comments to the dimension swapping stuff there. I can't remember the details. Maybe you can try to remove it and see what tests don't work any more? If all still works, leave it out. The test coverage is quite good, especially in the huge "full" model test.

In case we get too many problems with all that swapping, we might think about changing the channel order in fdeep from CHW to HWC completely, and thus match the Keras order. It was quite painful to get it working. And the reason for CHW over HWC is historical. The convolution implementation without matrix multiplication was faster with CHW. But currently there is no reason (except backward compatibility) to continue with HWC if it becomes too much of a burden.

**Comment by n-Guard** [comment]
I'm a little bit stuck at the moment so I appreciate any suggestion (@chammika-become / @chammika @Dobiasd).
One problem is: when we use `tensor3s.size()` as `seq_len`, then `model.predict_multi()` would not make sense anymore with` LSTMs`, since it takes every `tensor3` element inside `tensor3s` (one timestep) separately and feeds it to a separate model instance which also leads to separate outputs. What we need would be something like a `vector<tensor3s>`, so that the `LSTM` layer can process one whole `tensor3s` at once, right? 

On the other hand regarding `TimeDistributed`: 
having separate inputs and outputs per timestep is exactly what we need when using a `TimeDistributed` wrapper, this is what `model.predict_multi()` already does. 

**But**: when having a model that first has e.g. some `Conv` layers wrapped with a `TimeDistributed` layer followed by `LSTM` layers, then basically what we would need is a model split into two parts: 
`model.predict_multi()` for the `TimeDistributed(Conv())` layers up until the `LSTM` layer, and then use the complete output of `model.predict_multi()` as one sequence to feed one `LSTM` layer. (I'm not proposing to actually split the model like this, but I'm not quite sure what the best approach would be to integrate this logic into fdeep.)

**Comment by Dobiasd** [comment]
`predict_multi` is only a thin wrapper for running multiple completely independent forward passes in parallel. It takes `std::vector<tensor3s>` (aka `std::vector<std::vector<tensor>>`). If you have `predict` (taking `tensor3s` aka `std::vector<tensor3>`) working, you don't need to worry about `predict_multi`. It will be correct automatically.

After the discussions with chammika I was under the impression, that LSTM only needs 4 dimensions for its input (`[seq_len, channels, height, width]`). If instead it needs something 5-dimensional we have a problem.

**Comment by n-Guard** [comment]
> After the discussions with chammika I was under the impression, that LSTM only needs 4 dimensions for its input ([seq_len, channels, height, width]). If instead it needs something 5-dimensional we have a problem.

This is actually not quite right. An LSTM (and RNNs in Keras in general) only takes a 2-dim input (`[seq_len, features]`) when leaving out batch dimension. https://keras.io/layers/recurrent/
On statefulness: my understanding is when having no batch dimension our model will always be stateful since at no point on our forward pass our states will be reset to zero again. We could simulate `stateful=False` by introducing a parameter (e.g. `reset_after_x_timesteps`) though.

But say we want to classify some objects in a sequence of pictures (aka video), then it's a common use case to first have some Conv operations on every timestep (image) separately (i.e. with `TimeDistributed` wrapper), then flatten the output before the LSTM layer to (`[seq_len, features, 1, 1]`). So this is why it would be nice to have this fourth dimension because the input to the Conv layers would indeed be (`[seq_len, channels, height, width]`) but **not** to the LSTM layers.

This problem is actually a little bit detached from the LSTM implementation itself, since the input to the LSTM has to be flattened to two dimensions at some point anyway.

**Comment by Dobiasd** [comment]
> So this is why it would be nice to have this fourth dimension ...

Sorry, in case I'm totally off, but we **do** have 4 dimensions already (`typedef std::vector<tensor3> tensor3s;`).
- `predict` takes `std::vector<tensor3>` and returns `std::vector<tensor3>`, so input and output are 4-dimensional.
- `predict_multi` takes `std::vector<std::vector<tensor3>>` and returns `std::vector<std::vector<tensor3>>`, so input and output are 5-dimensional.

`predict_multi` should never be of any relevance when implementing a new layer type.

All [`layer`](https://github.com/Dobiasd/frugally-deep/blob/master/include/fdeep/layers/conv_2d_layer.hpp#L51)s (e.g. [`conv_2d_layer`](https://github.com/Dobiasd/frugally-deep/blob/master/include/fdeep/layers/layer.hpp#L53)) also take and return `std::vector<tensor3>` (aka `tensor3s`). In case of `conv_2d_layer` the outer dimension has to be `1`, but for others (like `concatenate_layer`) the outer dimension can be greater than `1`.

**Comment by chammika** [comment]
My idea was to use `[seq_len, features, 1, 1]` exactly as @n-Guard identified. My understanding was even though `width` `hight` is `1` it doesn't matter as `features` will always(?) need single dimension.
Please check out my attempt to make use of @n-Guard LSTM layer implementation and trying to integrate it to fdeep. https://github.com/chammika/frugally-deep/commit/cb869ee6562ec5ccf248256bf4d8bff832d0c5f8 

This is work in progress note the TODOs, sorry I am not good with Eigen, need to transform the serialized `float_vec`s to Eigen vectors. I uses the convention of other layer constructors which accepts `float_vec` s -the de-serialized yet un-shaped weights. Constructor is updated accordingly. Below can be use do generate a LSTM test model.

`generate_test_models.py recurrent recurrent_test.h5`
`convert_model.py recurrent_tests.h5 recurrent_test.json`

I am currently stuck at tests created by `convert_model.py`. it swaps input shape that I couldn't properly generate the input shape expected by the model.predict(). @n-Guard if you can get some ideas from this and improve your PR it would be great.

Note: my handles `chammika` (personal) and `chammika-become` (office) sorry for the confusion

**Comment by Dobiasd** [comment]
@gnosky10 Not yet, but it's [work in progress](https://github.com/Dobiasd/frugally-deep/pull/83). :-)

**Comment by Dobiasd** [comment]
Closed with the merge of [PR83](https://github.com/Dobiasd/frugally-deep/pull/83).

**Comment by JinhoPark-SNU** [comment]
Thank you. Great work! It would be wonderful to support Masking layers, because when one uses LSTM layers, he tends to use Masking layers also.

**Comment by Dobiasd** [comment]
@JinhoPark-SNU If I understand correctly, this would not just mean to add the masking layer itself to fdeep, but also to modify other layers that need to support it.

@n-Guard What do you think?

**Comment by Dobiasd** [comment]
Right. Up to now we only forward the raw tensors. It probably would result in some modifications in `get_layer_output`/`layer::get_output`/`node::get_output`.

The basic idea of these functions is to not push the data trough the model from front to end, but instead pull it out from the end. This "pull" then propagates through the computational graph up to the input layer(s).

One advantage of this is the following.

Consider we have such a graph (`A` is our only input layer, `H` is our only output layer):

```
         /-->C---->D
A---->B--                   -->H
         \-->E---->F--->G--/

```

Pushing from `A` would also invoke `C` and `D`. But actually computing these is not needed. Pulling from `H` solves this issue. The calculations in `C` and `D` will not be executed.

**Comment by Dobiasd** [comment]
Moved our conversation to a [new issue](https://github.com/Dobiasd/frugally-deep/issues/94). :)

**Comment by marco-monforte** [comment]
Hi guys,
First and foremost, thanks for the great job you're doing!

For the project I'm working on I need stateful LSTM (and a way to reset the internal cell state). Is this implemented in any of your functions? Unfortunately, my knowledge of C++ is limited and I don't feel I can help with it.

**Comment by Dobiasd** [comment]
@marco-monforte So you are looking for something like this in frugally-deep?

```cpp
const auto result_1 = model.predict(some_input_tensors);
model.reset_some_hidden_LSTM_state();
const auto result_2 = model.predict(some_other:input_tensors);
```

**Comment by marco-monforte** [comment]
@Dobiasd yes! And if I input two consecutive tensors, the hidden LSTM state is preserved from the first call to the second

**Comment by Dobiasd** [comment]
Currently something like that is not supported. [`fdeep::model::predict` is even const](https://github.com/Dobiasd/frugally-deep/blob/80c760d8ff15cc27ef76dde0b2a385e861fb47cd/include/fdeep/model.hpp#L25), meaning it is not intended to change internal state. What you are trying to achieve can not be done with a `TimeDistributed` layer, right?

**Comment by marco-monforte** [comment]
No, it's a different concept from the ``TimeDistributed`` layer.

When we run a prediction, the LSTM network output is built upon an internal state of the cells, obtained from the gates. What the model outputs, however, it's just the first information and the state is resetted after each prediction. In some cases, it's useful to keep this cell state intact so that the next prediction will depend also from the previous one, but then we have to manually control this resetting of the memory. It is something used in particular cases, also because of the "danger" associated to this memory preservation.

Thanks anyway, the library is really great! Hopefully someone will add this feature soon :)

**Comment by keithchugg** [comment]
@Dobiasd - what @marco-monforte is asking for is very useful.  A common case is that an RNN is trained with input:  (batch_size, sequence_length, features_dim) and stateful=False, return_sequence=True.  Then you can convert the trained model to another "streaming" keras model that has input: (1,1, features_dim) and stateful=True, return_sequence=False.  This streaming model then with an indefinite sequence length.  If you want to start over, you reset the states explicitly.

**Comment by Dobiasd** [comment]
Thanks for the explanation. Could you give a minimal Keras code example that does it?

**Comment by keithchugg** [comment]
Below is an example.  The [feature you added to the LSTM layer](https://github.com/Dobiasd/frugally-deep/commit/f45b2201d1dbe275e6a713412bb043fbb07dc77d) to set the state could be use to make this work.  The state would have to be returned and then sent back as an input for the next step.  It would be cleaner if the recursive layers had a "stateful" (remember state) and reset_state() options.  

Thanks for frugally-deep!  

```python 
import h5py
import keras
import numpy as np
from keras.layers import Input, Dense, GRU

##### generate toy data
train_seq_length = 4
feature_dim = 2
num_seqs = 8
x =  np.random.randint(0, high=2, size = (num_seqs * train_seq_length, feature_dim) )
x = np.sign( x - 0.5 )
y = np.sum( ( x == np.roll(x, 1, axis = 0) ), axis = 1 )
### y[n] = number of agreements between x[n], x[n-1]
x = x.reshape( (num_seqs, train_seq_length, feature_dim) )
y = y.reshape( (num_seqs, train_seq_length, 1) )


######  Define/Build/Train Training Model
training_in_shape = x.shape[1:]
training_in = Input(shape=training_in_shape)
# training_in = Input(batch_shape=(None,train_seq_length,feature_dim)) this works too
foo = GRU(4, return_sequences=True, stateful=False)(training_in)
training_pred = Dense(1)(foo)

training_model = keras.Model(inputs=training_in, outputs=training_pred)
training_model.compile(loss='mean_squared_error', optimizer='adam')
training_model.summary()

training_model.fit(x, y, batch_size=2, epochs=10)

##### define the streaming-infernece model
streaming_in = Input(batch_shape=(1,1,feature_dim))  ## stateful ==> needs batch_shape specified
foo = GRU(4, return_sequences=False, stateful=True )(streaming_in)
streaming_pred = Dense(1)(foo)
streaming_model = keras.Model(inputs=streaming_in, outputs=streaming_pred)

streaming_model.compile(loss='mean_squared_error', optimizer='adam')
streaming_model.summary()

##### copy the weights from trained model to streaming-inference model
training_model.save_weights('weights.hd5', overwrite=True)
streaming_model.load_weights('weights.hd5')

##### demo the behaivor
print('\n\n******the streaming-inference model can replicate the sequence-based trained model:\n')
for s in range(num_seqs):
    print(f'\n\nRunning Sequence {s} with STATE RESET:\n')
    in_seq = x[s].reshape( (1, train_seq_length, feature_dim) )
    seq_pred = training_model.predict(in_seq)
    seq_pred = seq_pred.reshape(train_seq_length)
    for n in range(train_seq_length):
        in_feature_vector = x[s][n].reshape(1,1,feature_dim)
        single_pred = streaming_model.predict(in_feature_vector)[0][0]
        print(f'Seq-model Prediction, Streaming-Model Prediction, difference [{n}]: {seq_pred[n] : 3.2f}, {single_pred : 3.2f}, {seq_pred[n] - single_pred: 3.2f}')
    streaming_model.reset_states()

print('\n\n******streaming-inference state needs reset between sequences to replicate sequence-based trained model:\n')
for s in range(num_seqs):
    print(f'\n\nRunning Sequence {s} with NO STATE RESET:\n')
    in_seq = x[s].reshape( (1, train_seq_length, feature_dim) )
    seq_pred = training_model.predict(in_seq)
    seq_pred = seq_pred.reshape(train_seq_length)
    for n in range(train_seq_length):
        in_feature_vector = x[s][n].reshape(1,1,feature_dim)
        single_pred = streaming_model.predict(in_feature_vector)[0][0]
        print(f'Seq-model Prediction, Streaming-Model Prediction, difference [{n}]: {seq_pred[n] : 3.2f}, {single_pred : 3.2f}, {seq_pred[n] - single_pred: 3.2f}')
    #### NO STATE RESET HERE: streaming model will treat multiples sequences as one long sequence, 
    #### so after first sequence, the streaming output will differ, difference will decay with time from start up as effect of intial state fades

for s in range(2):
    N = np.random.randint(1, 10)
    print(f'\n\n******streaming-inference can work on an sequences of indefinite length -- running length {N}:\n')
    for n in range(N):
        x_sample =  np.random.randint(0, high=2, size = ( 1, 1, feature_dim) )
        x_sample = np.sign( x_sample - 0.5 )
        single_pred = streaming_model.predict(x_sample)[0][0]
        print(f'Streaming-Model Prediction[{n}]:  {single_pred : 3.2f}')
```

**Comment by Dobiasd** [comment]
Thanks a lot. I'm still in the process of trying to understand it, but you seem to already do. Also, you write good code. Would you be interested in trying to implement it in frugally-deep in a PR?

I'd prefer the cleaner solution, you proposed. If I understand correctly, we would need to give up the `const`-ness of `model::predict`. But to still have this class thread-safe, we could use [`thread_local` storage](https://en.cppreference.com/w/cpp/language/storage_duration) for the mutating state. What do you think?

**Comment by keithchugg** [comment]
Regarding the code sample, the main point is that the training_model is trained using sequences and w/o statefulness, so each sequence is separate training sample and starts from the zero-state.  The streaming model is the same as the training model, except: (i) the input shape is just one time sample (sequence length 1) and (ii) it is stateful, meaning that each call picks up from where the last left off.  You need to use sequences of a fixed length to train, but you may want to run the trained model on a time series of indefinite length and the streaming model does that. 

I am happy to help with adding this feature to the recurrent layers, but I am just starting to familiarize myself with frugally-deep and am a C++ novice -- I don't really get all of the headerless and lamba stuff.  Despite you nice comment, I am not much of a programmer.  I understand the recurrent layer functionality and math pretty well though...

If the recurrent layer was a class in C++, then the feature could be implemented with (i) a private vector that is the state (in LSTMs, this is represented as two state vectors typically) (ii) a public function that would allow you to set/reset the state, and (iii) another private variable defining if the layer is stateful (ie., if the state is reset to 0 at each call or if it retains the previous state).  

I think what you are saying is that your current implementation does not allow for state in model::predict (for thread-safe reasons?) and that you could accomplish the vanilla-C++ approach above using the thread_local storage method.  I don't really understand any of that.. ;-) -- but yeah, if that accomplishes the same thing with the benefits you mention, great!

BTW, in keras, the reset_states() for a model resets the states of all recurrent layers and the state of a given layer can be set to a specific value using:

my_model.layers[i].reset_states()  # sets to zero
my_model.layers[i].reset_states(my_desired_state) # sets to my_desired_state

**Comment by keithchugg** [comment]
I looked at your code a little bit: lstm_impl, gru_impl.  The desired functionality can be achieved by passing in the state(s) and returning the state -- this is the "less clean" approach, but fits with your functional approach.  For the GRU, this is simple because the current_state is just the last output so you do not need to return anything more -- you just need to pass in initial_state_h.  Then, it would work like this (in sort of pseudo-C):

```C
state = initial_h;
for(n=0; n< N; ++n){
     y[n] = gru_predict( x[n], state);
     state = y[n];
}
```

For the LSTM, the state is broken in two: "h" is the output and "c" is an internal state.  So, you would have to return the state "c" at the end of the sequence to implement the streaming mode.  

The private variable approach seems cleaner to me, but it is less in line with the style of your code.  Although I am not familiar with many of the methods you are using in your code, it is very nice -- I can still tell what you are doing!

**Comment by Dobiasd** [comment]
> I am happy to help with adding this feature to the recurrent layers, but I am just starting to familiarize myself with frugally-deep and am a C++ novice -- I don't really get all of the headerless and lamba stuff. Despite you nice comment, I am not much of a programmer. BTW, if you have any good resources to get up to speed on the C++ approaches used in fd, please share (I am starting from ansi C and some C++).

Ah, OK, understood.
With "headerless" I guess you mean "header-only". :)
Regarding the usage of FunctionalPlus in frugally-deep's code, the [readme file of it](https://github.com/Dobiasd/FunctionalPlus#tutorial) might be a good starting point.

---

> If the recurrent layer was a class in C++, then the feature could be implemented with (i) a private vector that is the state (in LSTMs, this is represented as two state vectors typically) (ii) a public function that would allow you to set/reset the state, and (iii) another private variable defining if the layer is stateful (ie., if the state is reset to 0 at each call or if it retains the previous state).

> BTW, in keras, the reset_states() for a model resets the states of all recurrent layers and the state of a given layer can be set to a specific value using:
> my_model.layers[i].reset_states() # sets to zero
> my_model.layers[i].reset_states(my_desired_state) # sets to my_desired_state

Yes, it is a [class in C++](https://github.com/Dobiasd/frugally-deep/blob/c47e5ab7913f4f8235c1e7ba58cb422743589a99/include/fdeep/layers/lstm_layer.hpp).

Currently, the model class does not provide public access to the layers. Also, [it only stores them as base-class (`fdeep::layer`) pointers](https://github.com/Dobiasd/frugally-deep/blob/c47e5ab7913f4f8235c1e7ba58cb422743589a99/include/fdeep/layers/model_layer.hpp#L73), so it does not even know the type of any of the layers it consists of.

One way, that allows us to provide a supple way for the user of `fdeep::model` to set the LSTM's state, might be: We add a `reset_states` member function to `fdeep::layer` in general. For stateless layers (then everything except LSTM), this would just do nothing (or raise an exception if more appropriate). For the LSTM layer it would be implemented to do whatever it needs to do there. No new private `is_stateful` flag would be needed. The model itself would expose a new `reset_states` function, which might be better than exposing the layers itself completely.

Would `fdeep::model::reset_states();`, i.e. resetting all LSTM states in the model, suffice for your use-case?
Or would you need a way to only reset specifig layers, i.e., `fdeep::model::reset_states(std::size_t layer_index);`?
Or even `fdeep::model::reset_states(std::size_t layer_index, const whatever_type& new_state);` to set specific states instead of just empty everything?

---

> I think what you are saying is that your current implementation does not allow for state in model::predict (for thread-safe reasons?) and that you could accomplish the vanilla-C++ approach above using the thread_local storage method. I don't really understand any of that.. ;-) -- but yeah, if that accomplishes the same thing with the benefits you mention, great!

Yeah, basically the approach with the state variables in the LSTM-layer class would become threadsafe when we simply declare them as `thread_local`. :)

---

Once we are clear about the questions from above, maybe we could share the work like the following:
- I adjust the architecture, i.e., `fdeep::model` and `fdeep::layer` and add a `reset_states` interface.
- You then could take care of adding the actual functionality to the LSTM layer. A pull request would be a good place for that. Adding test cases if possible (`generate_test_models.py`) would be nice too.
- Once it's working (tests are green), we can take care of the thread-safety part.

**Comment by Dobiasd** [comment]
> I just did not understand how big of a deal it would be to have state in the layers.
I think no big deal at all, if we manage to encapsulate it nicely

> One question: you have initial states passed to the LSTM layer. I am not sure how these get set from the model object.

The same as Keras does. If an LSTM layer is invoked with only one input tensor, that means no initial states. If there are 3 input tensors, then the last two ones represent the initial state. This happens [here in the code](https://github.com/Dobiasd/frugally-deep/blob/c47e5ab7913f4f8235c1e7ba58cb422743589a99/include/fdeep/layers/lstm_layer.hpp#L58-L59).

> It could be cleaner to have reset_states() take optional arguments that are the state values to be set.

I thought "99% of applications" don't require that, and simply setting to all zeros is enough. Or did I misunderstand that?

---

Maybe we just start with that simple approach, and then see how it goes.

I suggest the following game plan.

Tobias:
- Open feature branch `stateful`.
- Add `layer::reset_states()` (dummy implementation that does nothing).
- Add `model::reset_states()` (calls every reset_states on every layer, also recursively for `model_layer`s).
- Call `model::reset_states()` after loading the model (to remove test-case state).
- Add example or overriding this function to `lstm_layer`.
- Add some `todo` comments to the layer.
- Commit and push.

Keith:
- Pull/checkout `stateful` branch
- Add new model `test_model_lstm_stateful` to `generate_test_models.py`.
- Commit, push, and open (WIP) pull request (so Tobias can review).
- Add the new test model to `CMakeLists.txt`.
- Add new `cpp` file for it in the `tests` directory.
- Add the new test model to `applications_performance.cpp` (for possible local debugging).
- Implement statefulness in the LSTM layer.
    - Use a `mutable` private member variable for this, so we can leave `model::predict` `const` for now, i.e. not break the existing API.
- Commit and push not only when completely done but also regularly in between.

**Comment by Dobiasd** [comment]
OK, done. I just created [the new branch](https://github.com/Dobiasd/frugally-deep/tree/stateful) and pushed [a commit](https://github.com/Dobiasd/frugally-deep/commit/506b8ec195d7893642ba2c5ab8308cafcaa832cc) with the architectural skeleton.

Let me know if you feel blocked at some point or something is missing, etc. :slightly_smiling_face:

**Comment by keithchugg** [comment]
OK.  Thanks.  I will start on it...

**Comment by keithchugg** [comment]
> > It could be cleaner to have reset_states() take optional arguments that are the state values to be set.
> 
> I thought "99% of applications" don't require that, and simply setting to all zeros is enough. Or did I misunderstand that?

Yes, I don't have a use to set the states to a specific value, I just though that this functionality and reset_states() are so similar, you may want to combine.  

The plan you posted is good.  I may be a little slow on the uptake, but I will work on it and let you know...

**Comment by Dobiasd** [comment]
> Yes, I don't have a use to set the states to a specific value, I just though that this functionality and reset_states() are so similar, you may want to combine.

Similar, but, maybe surprisingly, way more complex. Having something like `model::reset_states(std::size_t layer_index, const tensor5s& new_state);` is only straight forward to simple sequential models, because of two cases:
- Models can be nested.
- In one model (outer or one of the nested ones), the computational graph can have parallel branches.
Currently, it's not clear to me how we could come up with meaningful indexing. We might need to expose the whole model architecture (graph + nesting) to the public API so that all layers can be traversed or found by some other means in order to set the state of a particular layer. If possible, I'd like to avoid increasing the "surface" of the library that drastically.

**Comment by marco-monforte** [comment]
Hey guys! I've been finally able to test the stateful LSTM in my C++ code and it works amazingly! Thank you very much! Very appreciated :smiley:

**Comment by marco-monforte** [comment]
Hey guys! I'm sorry to open again this issue, but I'm working again on my old code with LSTMs and I'm not able to adjust these two lines of code to the new tensors definitions:

``` 
deep::tensor5s result = model_stateful->predict_stateful({fdeep::tensor5(fdeep::shape5(1, 1, 1, 1, 3), {inputX, inputY, inputZ})});
std::vector<float> vec = *result.front().as_vector();
```

Could you help me?

**Comment by Dobiasd** [comment]
I'm afk right now. What error message do you get?
Have you tried removing the *?

**Comment by Dobiasd** [comment]
The following two parts of the FAQ (together with the usage example in the readme) cover your use-case:
- https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#how-to-fill-an-fdeeptensor-with-values-eg-from-an-stdvectorfloat
- https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#how-to-convert-an-fdeeptensor-to-an-stdvectorfloat

**Comment by marco-monforte** [comment]
Thanks! Apparently, I've been able to compile by doing the following:
```        
fdeep::tensor t(fdeep::tensor_shape(1,1,1,1,3),0);
t.set(fdeep::tensor_pos(0,0,0,0,0), x);
t.set(fdeep::tensor_pos(0,0,0,0,1), y);
t.set(fdeep::tensor_pos(0,0,0,0,2), z);
auto result = model_stateful->predict_stateful({t});
std::vector<float> vec = result[0].to_vector();
```

but then I get this error at runtime:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  Invalid inputs shape.
The model takes [2(Nothing, Just 3)] but provided was: [5(1, 1, 1, 1, 3)]
```
I don't get the meaning of  [2(Nothing, Just 3)], while the second pair of brackets should be right, given my t tensor.

**Comment by Dobiasd** [comment]
Since your problem is not related to LSTM layers, I guess this is not the right place to discuss this. Also, since it might spam the other participants of this thread with notifications, that are of no interest to them.

Basically, you're not providing the right input shape for your model. Try to only provide sizes for the dimensions, that are actually used, i.e., the last two.

In case of further problems, please open a separate issue.
